File: es-5-6-optimization-translation/twitter.cs
using System;
using System.Collections.Generic;

public class Twitter
{
    private int count;
    private Dictionary<int, List<int[]>> tweetMap;
    private Dictionary<int, HashSet<int>> followMap;

    public Twitter()
    {
        this.count = 0;
        this.tweetMap = new Dictionary<int, List<int[]>>();
        this.followMap = new Dictionary<int, HashSet<int>>();
    }

    public void PostTweet(int userId, int tweetId)
    {
        if (!tweetMap.ContainsKey(userId))
        {
            tweetMap[userId] = new List<int[]>();
        }
        tweetMap[userId].Add(new int[] { count, tweetId });
        if (tweetMap[userId].Count > 10)
        {
            tweetMap[userId].RemoveAt(0);
        }
        count--;
    }

    public List<int> GetNewsFeed(int userId)
    {
        var res = new List<int>();
        var minHeap = new PriorityQueue<int[], int>(Comparer<int>.Default);

        if (!followMap.ContainsKey(userId))
        {
            followMap[userId] = new HashSet<int>();
        }
        followMap[userId].Add(userId);

        foreach (var followeeId in followMap[userId])
        {
            if (!tweetMap.ContainsKey(followeeId)) continue;

            var tweets = tweetMap[followeeId];
            int index = tweets.Count - 1;
            var tweet = tweets[index];
            minHeap.Enqueue(new int[] { tweet[0], tweet[1], followeeId, index - 1 }, tweet[0]);
        }

        while (minHeap.Count > 0 && res.Count < 10)
        {
            var top = minHeap.Dequeue();
            res.Add(top[1]);
            int nextIndex = top[3];
            if (nextIndex >= 0)
            {
                var tweets = tweetMap[top[2]];
                var nextTweet = tweets[nextIndex];
                minHeap.Enqueue(new int[] { nextTweet[0], nextTweet[1], top[2], nextIndex - 1 }, nextTweet[0]);
            }
        }

        return res;
    }

    public void Follow(int followerId, int followeeId)
    {
        if (!followMap.ContainsKey(followerId))
        {
            followMap[followerId] = new HashSet<int>();
        }
        followMap[followerId].Add(followeeId);
    }

    public void Unfollow(int followerId, int followeeId)
    {
        if (followMap.ContainsKey(followerId))
        {
            followMap[followerId].Remove(followeeId);
        }
    }
}

File: es-8-visual-to-code/flowchart-to-code/Twitter.java
import java.util.*;

public class Twitter {
    private static int timestamp = 0;

    // Tweet class to store tweet id and time
    private static class Tweet {
        int id;
        int time;
        public Tweet(int id, int time) {
            this.id = id;
            this.time = time;
        }
    }

    // userId -> list of tweets (most recent first)
    private Map<Integer, List<Tweet>> userTweets;
    // userId -> set of followees
    private Map<Integer, Set<Integer>> followers;

    public Twitter() {
        userTweets = new HashMap<>();
        followers = new HashMap<>();
    }

    // Publish a tweet
    public void postTweet(int userId, int tweetId) {
        userTweets.putIfAbsent(userId, new LinkedList<>());
        userTweets.get(userId).add(0, new Tweet(tweetId, timestamp++));
    }

    // View recent tweets (news feed)
    // Returns up to 10 most recent tweet ids in the user's news feed.
    // Each item in the news feed must be posted by users who the user followed or by the user themself.
    public List<Integer> getNewsFeed(int userId) {
        PriorityQueue<Tweet> pq = new PriorityQueue<>((a, b) -> b.time - a.time);

        // Add user's tweets
        if (userTweets.containsKey(userId)) {
            pq.addAll(userTweets.get(userId));
        }

        // Add followees' tweets
        Set<Integer> followees = followers.getOrDefault(userId, new HashSet<>());
        for (int followeeId : followees) {
            if (userTweets.containsKey(followeeId)) {
                pq.addAll(userTweets.get(followeeId));
            }
        }

        // Get top 10 tweets
        List<Integer> res = new ArrayList<>();
        int count = 0;
        while (!pq.isEmpty() && count < 10) {
            res.add(pq.poll().id);
            count++;
        }
        return res;
    }

    // Follow a user
    public void follow(int followerId, int followeeId) {
        if (followerId == followeeId) return; // Can't follow yourself
        followers.putIfAbsent(followerId, new HashSet<>());
        followers.get(followerId).add(followeeId);
    }

    // Unfollow a user
    public void unfollow(int followerId, int followeeId) {
        if (followerId == followeeId) return; // Can't unfollow yourself
        Set<Integer> followees = followers.get(followerId);
        if (followees != null) {
            followees.remove(followeeId);
        }
    }

    // Simple main to demo the flow
    public static void main(String[] args) {
        Twitter twitter = new Twitter();

        // Publish a tweet
        twitter.postTweet(1, 101); // User 1 tweets 101

        // User 2 follows user 1
        twitter.follow(2, 1);

        // User 2's news feed should have tweet 101
        System.out.println("User 2's feed: " + twitter.getNewsFeed(2)); // [101]

        // User 2 unfollows user 1
        twitter.unfollow(2, 1);

        // User 2's feed (should be empty)
        System.out.println("User 2's feed after unfollow: " + twitter.getNewsFeed(2)); // []
    }
}